downloadDirectory/src/main/java/org/reactome/release/downloadDirectory/GenerateGOAnnotationFile/BiologicalProcessAnnotationBuilder.java

```diff
@@ -0,0 +1,169 @@
+                }
+            } else {
+                logger.info("Invalid protein, skipping GO annotation");
+            }
+        }
+    }
+
+    /**
+     * Before creating GOA line for BP annotations, the reaction in question needs to be checked for the existance of a 'goBiologicalProcess' attribute. If there is none
+     * than the instance's 'hasEvent' referrals are checked for any.
+     * @param referenceEntityInst -- GKInstance, ReferenceEntity instance from protein instance.
+     * @param reactionInst -- GKInstance, parent reaction instance.
+     * @param taxonIdentifier -- String, CrossReference ID of protein's species.
+     * @throws Exception -- MySQLAdaptor exception.
+     */
+    private static void getGOBiologicalProcessLine(GKInstance referenceEntityInst, GKInstance reactionInst, String taxonIdentifier) throws Exception {
+        for (Map<String, String> biologicalProcessAccession : getGOBiologicalProcessAccessions(reactionInst, 0)) {
+            String goaLine = GOAGeneratorUtilities.generateGOALine(referenceEntityInst, GOAGeneratorConstants.BIOLOGICAL_PROCESS_LETTER, biologicalProcessAccession.get(GOAGeneratorConstants.ACCESSION_STRING), biologicalProcessAccession.get(GOAGeneratorConstants.EVENT_STRING),  GOAGeneratorConstants.TRACEABLE_AUTHOR_STATEMENT_CODE, taxonIdentifier);
+            GOAGeneratorUtilities.assignDateForGOALine(reactionInst, goaLine);
+        }
+    }
+
+    /**
+     * This method checks for a populated 'goBiologicalProcess' attribute in the incoming instance. If there are none and the max recursion has been reached,
+     * it's 'hasEvent' referral is checked for it. Once finding it, it returns the 'accession' and 'identifier' for each one, which will be used to generate a GOA line.
+     * @param eventInst -- GKInstance, Can be the original reaction instance, or, if it had no Biological Process accessions, its Event referrals.
+     * @param recursion -- Integer, Indicates number of times the method has been recursively called.
+     * @return -- 1 or more Maps containing the GO accession string and event instance it is associated with.
+     * @throws Exception -- MySQLAdaptor exception.
+     */
+    private static List<Map<String, String>> getGOBiologicalProcessAccessions(GKInstance eventInst, int recursion) throws Exception {
+        List<Map<String, String>> goBiologicalProcessAccessions = new ArrayList<>();
+        if (recursion <= MAX_RECURSION_LEVEL) {
+            Collection<GKInstance> goBiologicalProcessInstances = eventInst.getAttributeValuesList(ReactomeJavaConstants.goBiologicalProcess);
+            if (goBiologicalProcessInstances.size() > 0) {
+                for (GKInstance goBiologicalProcessInst : goBiologicalProcessInstances) {
+                    if (!GOAGeneratorUtilities.proteinBindingAnnotation(goBiologicalProcessInst.getAttributeValue(ReactomeJavaConstants.accession))) {
```

Comment:
For clarity, could this be extracted to a helper method named something like `doesNotHaveProteinBindingAnnotation(goBiologicalProcessInst)`?